### ISSUE TITLE
Code splitting

### ISSUE THREAD

**Issue by anwerso** [catberry/catberry#209]
Let's say we have some heavy components, which used rather rarely. Sure we want to load/use such components only when they needed. Is there a way to split browser bundle to chunks (modules) which will be loaded later on-demand?

**Comment by pragmadash** [comment]
There is no way to do that in Catberry so far. But I'm planning to work on it in the future. Let's leave this issue as a feature request.

**Comment by markuplab** [comment]
Now, you can use [Catberry Plugin API](http://catberry.org/documentation#browserify-transformation-api) with [Browserify Factor Bundle](https://github.com/substack/factor-bundle).

**Comment by pragmadash** [comment]
@markuplab @anwerso, guys I would appreciate it if you make a plugin for that. Or at least a proof of concept and we will add it to the Catberry organization on Github.

**Comment by anwerso** [comment]
@markuplab, [partition-bundle](https://www.npmjs.com/package/partition-bundle) looks more apropriate for on-demand loading, but thanks for pointing to a plugin API.

@pragmadash, we just started to look to "the next big thing" for our projects. Maybe we will write some sort of tests/proof of concept later, not sure when though.

I also want to explore possibility of using [CSS modues](https://github.com/css-modules/css-modules) workflow with Catberry, but it's for another issue.

**Comment by pragmader** [comment]
We can start with splitting bundle into 2 parts:

* everything from `node_modules` directory
* app code

That would be simpler to implement and gives valuable benefit if you you have feature releases in your project.

**Comment by pragmader** [comment]
# API proposal

* create `.catberry` directory in the project root
* move there the tmp file `__BrowserBundle.js`
* also, on first build put there files with checksums or last modification dates of `node_modules`, components and stores directories
* when the next build occurs it checks if `node_modules`, components or stores directories were changed and if `node_modules.js` and `app.js` files exist in the build directory
* run build if needed or just skip the whole process

In this case the build process would be fast as speed of light.

**Comment by pragmader** [comment]
I'm wondering, what if we finally [switch to webpack](https://github.com/catberry/catberry/issues/271), would it solve this as well?

**Comment by reenko** [comment]
May be will be load configs from other file (not bundle.js)? Because, we will be generate one bundle.js for any configs/evn.

**Comment by pragmader** [comment]
@reenko you can easily do it without any change in the framework:
* use a separate JS-file as a configuration, this file defines a global config variable for a browser environment
* don't `require` this file anywhere
* include a `<script>` with the config file before `bundle.js`
* use the global config variable in `browser.js` like `const cat = catberry.create(glbalConfig);`
* use [`process.env`](https://nodejs.org/api/process.html#process_process_env) for `server.js` configuration.

In this case you can build, for instance, a container image for all environments and set the environment variables on start. Also, you would have to generate the browser configuration before starting the process, the [catberry-homepage script](https://github.com/catberry/catberry-homepage/blob/develop/Makefile#L16-L31) can be used for that.

**Comment by pragmader** [comment]
Let's start with a simple step: split up `bundle.js` into `app.js` and `vendor.js` by filenames (if it contains `node_modules` in its filename it goes to `vendor.js`. It would allow to cache `vendor.js` somewhere on CDN and make changes in a smaller `app.js` script only making users to download `app.js` only which is much better.

**Comment by pragmader** [comment]
Implemented the bundle split into: `app.js` and `externals.js`.
No lazy loading for now, I doubt it would be efficient anyway. Closing.
